src/software/util/parameter/config.hpp

```diff
@@ -0,0 +1,242 @@
+            "we_have_no_chill", "should_we_still_have_namespaces", false);
+
+        internal_param_list = {we_have_no_chill_param, we_suck_param};
+    }
+
+    ~FriendlyEvalConfig() {}
+
+    const std::shared_ptr<Parameter<bool>> weSuck()
+    {
+        return we_suck_param;
+    }
+
+
+    const std::shared_ptr<Parameter<bool>> weHaveNoChill()
+    {
+        return we_have_no_chill_param;
+    }
+
+    std::string name()
+    {
+        return "TestConfigA";
+    }
+
+    ParameterList& getMutableParameterList()
+    {
+        return internal_param_list;
+    }
+
+    const ParameterList& getParameterList()
+    {
+        return internal_param_list;
+    }
+
+   private:
+    ParameterList internal_param_list;
+    std::shared_ptr<Parameter<bool>> we_suck_param;
+    std::shared_ptr<Parameter<bool>> we_have_no_chill_param;
+};
+
+
+class EnemyEvalConfig : public MutableConfig
+{
+   public:
+    EnemyEvalConfig()
+    {
+        // The parameters are initialized in the constructor
+        enemy_team_sucks_param = std::make_shared<Parameter<bool>>(
+            "enemy_team_sucks", "should_we_still_have_namespaces", true);
+        enemy_team_has_no_chill_param = std::make_shared<Parameter<bool>>(
+            "enemy_team_has_no_chill", "should_we_still_have_namespaces", true);
+
+        internal_param_list = {enemy_team_sucks_param, enemy_team_has_no_chill_param};
+    }
+
+    ~EnemyEvalConfig() {}
+
+    const std::shared_ptr<Parameter<bool>> enemyTeamSucks()
+    {
+        return enemy_team_sucks_param;
+    }
+
+    const std::shared_ptr<Parameter<bool>> enemyTeamHasNoChill()
+    {
+        return enemy_team_has_no_chill_param;
+    }
+
+    std::string name()
+    {
+        return "enemyEvalConfig";
+    }
+
+    ParameterList& getMutableParameterList()
+    {
+        return internal_param_list;
+    }
+
+    const ParameterList& getParameterList()
+    {
+        return internal_param_list;
+    }
+
+   private:
+    ParameterList internal_param_list;
+    std::shared_ptr<Parameter<bool>> enemy_team_sucks_param;
+    std::shared_ptr<Parameter<bool>> enemy_team_has_no_chill_param;
+};
+
+
+class GlobalQualityConfig : public MutableConfig
+{
+   public:
+    GlobalQualityConfig()
+    {
+        // The parameters are initialized in the constructor
+        everyone_sucks_param = std::make_shared<Parameter<bool>>(
+            "everyone_sucks", "should_we_still_have_namespaces", true);
+
+        friendly_eval_config = std::make_shared<FriendlyEvalConfig>();
+        enemy_eval_config    = std::make_shared<EnemyEvalConfig>();
+
+        internal_param_list = {everyone_sucks_param, friendly_eval_config,
+                               enemy_eval_config};
+    }
+
+    const std::shared_ptr<Parameter<bool>> everyoneSucks()
+    {
+        return everyone_sucks_param;
+    }
+
+    const std::shared_ptr<FriendlyEvalConfig> getFriendlyEvalConfig()
+    {
+        return friendly_eval_config;
```

Comment:
Just checking: Should we be able to call getMutableParameterList on friendly_eval_config? This function is marked const so should we be able to modify friendly_eval_config from here? Is it possible to return the immutable version of friendly_eval_config?

Reply:
So the idea is we create a `GlobalQualityConfig` somewhere HL, and then pass in const ref to all the plays/tactics/functions that need parameters. The visualizer would create a `GlobalQualityConfig` and get call `getMutableParameterList()` and would recursively visit and update the parameters.

But yeah I forgot to add the mutable functions like `getMutableFriendlyEvalConfig()` and `getMutableEveryoneSucks()` for testing purposes (and for special cases in the visualizer where it looks for params like `runAI` and does specific things). Since we would use `const` everywhere except for visualizer (for now), we could enforce that nobody else changes the params.

So yeah it totally is possible to add that without exposing too much to AI if we setup `const` correctly, ill tweak the python script to reflect this

Reply:
That's a good point. I think we might need to end up having two classes for each config, an immutable version and a mutable version that extends the immutable version. So we'd have `GlobalQualityConfig : public Config` and `GlobalQualityMutableConfig : public GlobalQualityConfig, MutableConfig`

Reply:
@akhilveeraghanta - if we pass a const ref to the config, I don't think we'll be able to call functions that return something like `const std::shared_ptr<Parameter<bool>>`, it would have to be `const std::shared_ptr<const Parameter<bool>>`.

Reply:
(otherwise you could call `global_quality_config.everyoneSucks()->setValue(false)`)

Reply:
Moving conversation to slack.........

Reply:
completely overlooked this, really good catch, I've updated it to be immutable using `std::const_cast_pointer` which creates a second shared pointer while keeping the pointee the same and casts to a `const` version of itself. 

ref: http://www.cplusplus.com/reference/memory/const_pointer_cast

Reply:
Oh nice! That's super cool. Just to double check, we now can't mutate variables unless we get the mutable iterator?

Reply:
yup, we cannot, this feature is pretty sweet

Reply:
cool cool, what if we want to mutate a specific parameter (say in some "play" that automatically calibrates physical constants). No worries if we don't support this in the initial revision, but might be something we'll need in the future.

Reply:
If we need this, we would have to separate them into `AIConfigured` and `visualizerConfigurable` for example to specify which ones we set and which ones the AI will set (assuming we don't want AI/Visualizer setting the same params) and then we could update the play that tunes params to get a mutable `AIConfigured`  instance (which the visuzlier will get an immutable version of to just inspect the values). Is this along the lines of what you are thinking?

We are going to be adding/update a lot of plays and tactics, might as well support this feature if its immediately useful.

Reply:
Hmm, I wasn't particuarly thinking of splitting them based on AI visualizer. This is probably easier discussed in person though, so let's hold off until tomorrow?